File: Platformer/scripts/Saw.jl
module SawModule
    using JulGame

    mutable struct Saw <: Script
        animator::AnimatorModule.Animator
        endingY::Float64
        isMovingUp::Bool
        rotation::Float64
        parent::JulGame.EntityModule.Entity
        sound::SoundSourceModule.SoundSource
        speed::Number
        startingY::Float64

        function Saw()
            this = new()

            this.endingY = 3
            this.isMovingUp = false
            this.rotation = 0.0
            this.speed = 5.0
            this.startingY = 0

            return this
        end
    end

    function JulGame.update(this::Saw, deltaTime)
        this.rotation += 5.0
        this.parent.sprite.rotation =  this.rotation % 360.0

        if this.parent.transform.position.y >= this.startingY && !this.isMovingUp
            this.isMovingUp = true
        elseif this.parent.transform.position.y <= this.endingY && this.isMovingUp
            this.isMovingUp = false
        end

        if this.isMovingUp
            this.parent.transform.position = Vector2f(this.parent.transform.position.x, this.parent.transform.position.y - this.speed*deltaTime)
        else
            this.parent.transform.position = Vector2f(this.parent.transform.position.x, this.parent.transform.position.y + this.speed*deltaTime)
        end
    end
end # module

File: Platformer/ts/_generated/scripts/Saw.ts
import { registerScript } from "julgame/src/engine/runtime/scriptRegistry";


    // using JulGame

    export class Saw {
        animator: Animator
        endingY: number
        isMovingUp: boolean
        rotation: number
        parent: Entity
        sound: SoundSource
        speed: Number
        startingY: number

        constructor() {
            

            this.endingY = 3.0
            this.isMovingUp = false
            this.rotation = 0.0
            this.speed = 5.0
            this.startingY = 0.0

        }
    }

    function JulGame_update_Saw(self: Saw, deltaTime) {
        self.rotation += 5.0
        self.parent.sprite.rotation =  self.rotation % 360.0

        if (self.parent.transform.position.y >= self.startingY && !self.isMovingUp) {
            self.isMovingUp = true
        } else if (self.parent.transform.position.y <= self.endingY && self.isMovingUp) {
            self.isMovingUp = false
        }

        if (self.isMovingUp) {
            self.parent.transform.position = {x: self.parent.transform.position.x, y: self.parent.transform.position.y - self.speed*deltaTime}
        } else {
            self.parent.transform.position = {x: self.parent.transform.position.x, y: self.parent.transform.position.y + self.speed*deltaTime}
        }
    }
registerScript("Saw", {
    create: () => new Saw(),
    update: JulGame_update_Saw,
});
